caffe2/core/dispatch/DispatchTable.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include "caffe2/utils/flat_hash_map/flat_hash_map.h"
+#include "caffe2/utils/Metaprogramming.h"
+#include "caffe2/core/dispatch/OpSchema.h"
+
+#include <type_traits>
+#include <array>
+#include <unordered_map>
+#include <iostream>
+#include <mutex>
+
+namespace c10 {
+
+namespace details {
+
+/// Internal implementation of the operator as a thread-safe hash table.
+template<class Key>
+class ThreadsafeOperatorTable_ final {
+public:
+    // TODO The current implementation below does not have the correct correctness characteristics
+    // which we need.  It's worth spelling out exactly what we need:
+    //
+    //  - We need LOCK FREE read access to the table (as per the performance benchmark
+    //    at https://fb.quip.com/hvz3AGnx8MQ8
+    //
+    //  - We need to support writes which are possibly concurrent with reads, occurring when
+    //    a dynamic library is loaded or unloaded.
+    //
+    //  - We probably can require that dynamic library loads/unloads be synchronized (so
+    //    there are never two concurrent loads.)
+
+    template<class Key_>
+    void emplace(Key_&& key, void* value) {
+      // TODO Locking
+      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+      auto result = map_.emplace(std::forward<Key>(key), value);
+      if (!result.second) {
+        throw std::logic_error("Tried to register conflicting operators to the dispatcher.");
+      }
+    }
+
+    void erase(const Key& key) {
+      // TODO Locking
+      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+      size_t num_removed = map_.erase(key);
+      assert(num_removed <= 1); //This is not a multi-map
+      if (num_removed == 0) {
+        throw std::logic_error("Tried to deregister an operator that isn't registered.");
+      }
+    }
+
+    void* lookup(const Key& key) const {
+      // TODO (lock needed but slow perf. Find better way)
+      // std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+      auto found = map_.find(key);
+      if (found == map_.end()) {
+        return nullptr;
+      } else {
+        return found->second;
+      }
+    }
+
+private:
+    ska::flat_hash_map<Key, void*> map_;
+    // TODO Figure out how to get fast locking in C++11 (use boost::shared_timed_mutex? folly::SharedMutex?)
```

Comment:
Actually, if we are not having much collisions accessing the hash map, a simple lock usually is very quick, and as initial step we probably do not need to worry about hyperoptimizing this, and would be fine to do a simple std::mutex first.

Reply:
I tested it with `std::shared_timed_mutex` (variant of `std::mutex` for fast reading) and it was horribly slow. Without the mutex, my benchmarks weren't able to see any perf overhead of the dispatcher. With the mutex, it was noticeable.

I'm not sure what the best way to go forward here is, I can either enable it and allow the perf regression, or disable it and it will crash in the (unlikely) case that two kernel libraries are loaded and try to register kernels at exactly the same time.

In any case, there will be a high-pri follow-up item to fix this and introduce a fast mutex.

Reply:
For loading libraries I think it's fine to be non-thread-safe towards reads. But you might have issue if you start registering new entries on cache misses. But we're not there yet so let's allow it to crash.